lib/webapp.js

```diff
@@ -917,38 +923,98 @@ WebServiceHandle.prototype = {
           options.body = json;
         }
       }
-      if (options.zluxLoopbackSecret) {
-        //TODO use secret in a crypto scheme
-        headers[ZLUX_LOOPBACK_HEADER] = options.zluxLoopbackSecret;
-      }
       //console.log("headers: ", headers)
       if (Object.getOwnPropertyNames(headers).length > 0) {
         requestOptions.headers = headers;
       }
-      let httpOrHttps = this.environment.loopbackConfig.isHttps ? https : http;
-      const request = httpOrHttps.request(requestOptions, (response) => {
-        var chunks = [];
-        response.on('data',(chunk)=> {
-          utilLog.debug('ZWED0194I'); //utilLog.debug('Callservice: Data received');
-          chunks.push(chunk);
-        });
-        response.on('end',() => {
-          utilLog.debug('ZWED0195I'); //utilLog.debug('Callservice: Service call completed.');
-          response.body = Buffer.concat(chunks).toString();
-          resolve(response);
-        });
+
+      let httpOrHttps;
+      if (this.isAgentService) {
+        requestOptions.hostname = this.environment.agentRequestOptions.host;
+        requestOptions.port = this.environment.agentRequestOptions.port;
+        requestOptions.protocol = this.environment.agentRequestOptions.protocol;
+        requestOptions.rejectUnauthorized = this.environment.agentRequestOptions.rejectUnauthorized;
+        if (this.environment.agentRequestOptions.apimlPrefix) {
+          requestOptions.path = this.environment.agentRequestOptions.apimlPrefix + requestOptions.path;
+        }
+        httpOrHttps = requestOptions.protocol == 'http:' ? http : https;
+        routingLog.debug(`Call agent host=%s path=%s`,requestOptions.hostname, requestOptions.path);
+      } else if (options.internalRouting) {
+        routingLog.debug(`Call internally path=%s`, requestOptions.path);
+        //clone done to prevent original request alteration, but shallow so hope attributes arent altered
+        let req = Object.assign({},originalRequest);
+        let res = Object.assign({},originalRes);
+
+        req.url = req.originalUrl = requestOptions.path;
+        req.method = requestOptions.method;
+        req.headers = requestOptions.headers;
+        req.rawHeaders=[];
+        let keys = Object.keys(req.headers);
+        for (let i = 0; i < keys.length; i++) {
+          let k = keys[i];
+          let v = req.headers[keys[i]];
+          req.rawHeaders.push(k);
```

Comment:
I am curious why do we need to set `rawHeaders`?

Reply:
I'm not sure if it is needed, but I just tried to mimic everything in the request as much as possible to avoid middleware breaking.